initial proposal revamp based on existing prototype (#7)
aswinayyolath/proposals: 095-stretch-cluster.md

# Stretch Kafka cluster

The Strimzi Kafka operator currently manages Kafka clusters within a single Kubernetes cluster.
This proposal aims to extend this support to stretch Kafka clusters, where brokers and controllers of a single Kafka cluster are distributed across multiple Kubernetes clusters.

## Current situation

At present, the availability of Strimzi-managed Kafka clusters is limited by the availability of the underlying Kubernetes cluster.
If a Kubernetes cluster experiences an outage, the entire Kafka cluster becomes unavailable, disrupting all connected Kafka clients.

## Motivation

> A stretch Kafka cluster allows Kafka nodes to be distributed across multiple Kubernetes clusters. This approach also facilitates maby valuable use cases, such as:

Reviewer: ```suggestion
A stretch Kafka cluster allows Kafka nodes to be distributed across multiple Kubernetes clusters. This approach also facilitates many valuable use cases, such as:
```

This necessitates careful consideration of network architecture and associated costs.

- **Latency Requirements**: Stretch Kafka clusters are best suited for environments with low-latency network connections between the Kubernetes clusters.
High latency can adversely affect the performance and synchronization of Kafka nodes, potentially leading to delays or errors in replication and client communication.
Defining the minimal acceptable latency between clusters is crucial to ensure optimal performance.

## Proposal

This proposal seeks to enhance the Strimzi Kafka operator to support stretch Kafka clusters, distributing brokers and controllers across multiple Kubernetes clusters.
The intent is to focus on high-availability of the data plane.
The proposal outlines high-level topology and design concepts for such deployments, with a plan to incrementally include finer design and implementation details for various aspects.

### Prerequisites

- **Multiple Kubernetes clusters**: Stretch Kafka clusters will require multiple Kubernetes clusters.

> Ideally, an odd number of clusters (at least three) will be needed to maintain Kafka controller quorum and tolerate a Kubernetes cluster outage.

Reviewer: ```suggestion
To ensure Kafka controller quorum in the event of a Kubernetes cluster outage, it is recommended to distribute controllers across at least three Kubernetes clusters. In a two cluster setup, there is a risk of losing quorum if the wrong cluster fails, which can lead to Kafka becoming unavailable.
```

- **Low Latency**: Kafka clusters should be deployed in environments that allow low-latency communication between Kafka brokers and controllers.
Stretch Kafka clusters should be deployed in environments such as data centers or availability zones within a single region, and not across distant regions where high latency could impair performance.

- **KRaft**: As Kafka and Strimzi transition towards KRaft-based clusters, this proposal focuses exclusively on enabling stretch deployments for KRaft-based Kafka clusters.
While Zookeeper-based deployments are still supported, they are outside the scope of this proposal.

- **A supported cloud native network technology**: To enable networking between Kubernetes clusters currently requires an additional technology stack.
The remainder of this proposal and the prototype detailed within assumes the use of [Submariner](https://submariner.io/), although there is potential for [Cilium](https://cilium.io/) support in a future revision.

> The [Istio](https://istio.io/) service mesh has some experimental support for [Kubernetes Multi-Cluster Services (MCS)](https://multicluster.sigs.k8s.io/concepts/multicluster-services-api/) that might also warrant further research.

Reviewer: I think this section could be elaborated a bit more and we should use a well-known term for describing what we need instead of creating our own term like `cross-cluster networking technology`. It would help if we could describe the capabilities, behaviour or features that we use form these technologies. 

CNCF talk about `cloud native network` as a category where both submariner and Cilium belong - https://landscape.cncf.io/guide#runtime--cloud-native-network. Cloud native networking uses software for controlling, inspecting and modifying data flows, making it easier to manage, secure and isolate connections between containers. 

They go on to describe specific capabilities using these words, but need one of you to look closely and include if these are in line with what we are leveraging:
- `In some cases you may want to extend your container networks and network policies such as firewall and access rules to allow an app to connect to virtual machines or services running outside the container network.` 
- `These tools create a virtual network on top of existing networks specifically for apps to communicate, referred to as an overlay network.`

I also like the use of [Kubernetes Multi-Cluster Services (MCS)](https://multicluster.sigs.k8s.io/concepts/multicluster-services-api/) as that is an open standard rather than a solution specific to one tool. If the changes we have made to the operator are relying on this then we could say that a cloud native network with support for Kubernetes multi-cluster services must be setup. This link shows a number of provides who implement this: https://multicluster.sigs.k8s.io/guides/#implementation-status.

Reviewer 2: Not all tools follows/use MCS. Submariner does, but Cilium doesn't

Author: I'm intending to push a commit that just talks about MCS for this revision.
@aswinayyolath once I've done that, can I ask you to revise given your recent success with Cilium and add details of the aspects of that 'Cloud Native Network' that are used to achieve a stretch cluster with our prototype.

> The [Istio](https://istio.io/) service mesh has some experimental support for [Kubernetes Multi-Cluster Services (MCS)](https://multicluster.sigs.k8s.io/concepts/multicluster-services-api/) that might also warrant further research.

Reviewer 2: ```suggestion
```

### Design

#### Topology of a stretch cluster

![Stretch cluster topology](./images/095-stretch-cluster-topology.png)

The diagram illustrates a topology comprising of three Kubernetes clusters.

One of these clusters is designated as the "Central cluster", while any additional clusters are considered "remote".

> The central cluster acts as the control plane where a user will create their Kafka cluster and Kafka node pool definitions (Custom Resources).

Reviewer: ```suggestion
The central cluster acts as the control plane where a user will create all the custom resources for the Kafka cluster - Kafka, KafkaNodePool, KafkaUser, KafkaTopic etc.
```

> A Kafka node pool definition can be configured to run on any of the Kubernetes clusters, including the central cluster.

Reviewer: ```suggestion
A Kafka node pool definition can be configured to specify a Kubernetes cluster (central cluster or one of the member clusters) as the deployment target.
The operator on the central cluster is responsible for creating all necessary resources (including StrimziPodSet resources) for the node pool on that specified Kubernetes cluster. 
```

> Operators deployed to remote clusters are only responsible for reconciling Strimzi pod set resources that are created remotely by the operator running in the central cluster.

Reviewer: ```suggestion
Operators deployed to remote clusters are only responsible for reconciling `StrimziPodSet` resources that are created remotely by the operator running in the central cluster.
```

This approach will allow users to manage the definition of their stretch Kafka cluster in a single location.

### Prototype

A working prototype can be deployed based upon Submariner technology using the steps outlined in a [draft README](https://aswinayyolath.github.io/stretch-kafka-docs/) that is being iteratively revised.
The following sections will describe the key aspects of the prototype highlighting areas that would benefit from community feedback and input.

#### User configuration of the cross-cluster Kubernetes network

> A cross-cluster network must be configured manually as a pre-requisite using a supported network technology.

Reviewer: cross-cluster Kubernetes network --> see previous comment

Reviewer 2: I think the standard term used in networking and K8s networking ecosystems for what Submariner and Cilium do is "inter-cluster networking" or "multi-cluster networking". I think we used use `multi-cluster networking` where ever possible

Author: The prototype is using Multi-Cluster Service features of Submariner, so I think we should stick with that. We'll need some extra words when we can add more about Cilium.

> The existing prototype uses Submariner where a broker must be deployed to one of the Kubernetes clusters involved and each cluster must be joined to that broker.
> This configuration must be performed by the user prior to deployment of Strimzi cluster operators.
> 
> During this process, a unique identifier is set for each Kubernetes cluster joined to the Submariner network.
> This identifier is used by the prototype central operator to build:

Reviewer: ```suggestion
The prototype uses Submariner to setup a cloud native network across multiple Kubernetes clusters. This configuration must be performed by the user prior to deploying Strimzi cluster operators.

Each Kubernetes cluster within the cloud native network is assigned a unique identifier.
This identifier is used by the prototype central operator to build:
```

1. Valid broker and controller service endpoints values for `advertised.listeners` and `controller.quorum.voters` within the appropriate ConfigMap resources.
2. Broker certificates with appropriate Subject Alternative Names.
The prototype currently expects this identifier to be supplied by the user as a label value on the Kafka node pool resource:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  ...
  labels:
    ...

>     strimzi.io/submariner-cluster-id: "cluster1"

Reviewer: ```suggestion
    strimzi.io/cluster-network-id: "cluster1"
```

I feel we should keep this generic or call out that we'll use the same field for all tools.

3. The existing Strimzi network policies block cross-cluster network traffic so the prototype requires the following environment variable to be set on the central operator deployment:

```yaml
- name: STRIMZI_NETWORK_POLICY_GENERATION
  value: 'false'
```

4. Submariner [requires an adjustment](https://aswinayyolath.github.io/stretch-kafka-docs/Deploying-Strimzi-in-Stretch-Mode/#updating-clusterrole-for-submariner) to the `ClusterRole` for the Strimzi cluster operator in the central Kubernetes cluster.

5. The prototype operator currently creates new Kubernetes client instances at the required stage of reconciliation for the range of Kubernetes resources involved.
This approach was quick to implement, but results in duplicate code.
A design for better management of Kubernetes clients for remote clusters is required and is dependent upon community input.

**All aspects of user configuration described above would benefit from further discussion within the community.**

## Additional considerations and reference information

#### Example of multi-cluster advertised.listener and controller.quorum.voters

> The Operator will set advertised.listener and controller.quorum.voters configurations as follows:

Reviewer 2: ```suggestion
The Operator configures advertised.listeners and controller.quorum.voters to support cross-cluster communication in a stretch Kafka deployment.
```

Reviewer 2: Now add below stuff under this

### Default advertised.listeners Format in Strimzi

In a standard Strimzi Kafka cluster (single-cluster setup), the `advertised.listeners` configuration follows this format

```yaml
advertised.listeners=REPLICATION-9091://<broker-pod-name>.<broker-service-name>.<namespace>.svc:9091,
PLAIN-9092://<broker-pod-name>.<broker-service-name>.<namespace>.svc:9092,
TLS-9093://<broker-pod-name>.<broker-service-name>.<namespace>.svc:9093
```

### Updated Format for Multi-Cluster (Stretch) Kafka

When Kafka is deployed in stretch mode, the Operator modifies advertised.listeners to include a cross-cluster identifier (e.g., Submariner or Cilium cluster-id). The updated format is

```yaml
advertised.listeners=REPLICATION-9091://<broker-pod-name>.<cross-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9091,
PLAIN-9092://<broker-pod-name>.<cross-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9092,
TLS-9093://<broker-pod-name>.<cross-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9093
```

The `<cross-cluster-id>` (Submariner or Cilium cluster-id) is dynamically retrieved from the KafkaNodePool label. This allows brokers in different Kubernetes clusters to discover and communicate with each other seamlessly.

### Updated controller.quorum.voters Format

Similarly, `controller.quorum.voters` includes all Kafka controllers running across multiple Kubernetes clusters. Each controller entry uses the cross-cluster DNS format to enable communication across clusters

```yaml
controller.quorum.voters=<controller-id>@<controller-pod-name>.<cross-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9090, .....
```

This ensures that controller nodes across multiple clusters can participate in leader elections and maintain cluster consistency.

> ```

Reviewer 2: ```suggestion
```

> advertised.listeners=REPLICATION-9091://my-cluster-broker-0.cluster1.my-cluster-kafka-brokers.strimzi.svc.clusterset.local:9091,PLAIN-9092://my-cluster-broker-0.cluster1.my-cluster-kafka-brokers.strimzi.svc:9092,TLS-9093://my-cluster-broker-0.cluster1.my-cluster-kafka-brokers.strimzi.svc:9093

Reviewer 2: ```suggestion
```

> controller.quorum.voters=101@my-cluster-controller-101.cluster2.my-cluster-kafka-brokers.strimzi.svc.clusterset.local:9090,1@my-cluster-controller-1.cluster1.my-cluster-kafka-brokers.strimzi.svc.clusterset.local:9090

Reviewer 2: ```suggestion
```

> ```

Reviewer 2: ```suggestion
```

These updates ensure brokers and controllers can be discovered and communicate across clusters without relying on traditional external access methods.

#### Resource cleanup on remote Kubernetes clusters

> As some of the Kubernetes resources will be created on a remote cluster, we will not be able to use standard Kubernetes approaches for deleting resources based on owner references.

Reviewer 2: ```suggestion
Currently, resources created in the remote clusters do not have OwnerReferences. The main reason is that the Kafka and KafkaNodePool CRs exist only in the central cluster. If resources in remote clusters were to reference them as owners, they would be immediately deleted by Kubernetes' garbage collection mechanism, since their owners don’t exist in the remote clusters.

Even if cross-cluster ownership were possible, it would introduce another issue: if the central cluster were to go down, any resources in remote clusters with OwnerReferences pointing to central cluster resources would also be deleted automatically. This is not the intended behavior, as we want remote resources to remain operational even if the central cluster becomes temporarily unavailable.

As of now, we have not yet implemented a mechanism for controlled cleanup of remote resources when a user deliberately deletes the Kafka and KafkaNodePool CRs from the central cluster. Currently, when a user deletes these CRs, all related resources in the central cluster are removed, but remote cluster resources remain. Finding a smart way to handle this cleanup while avoiding unintended deletions is an open challenge.
```

> The operator will need to delete remote resources explicitly when the owning resource is deleted.

Reviewer 2: ```suggestion
```

> - The exact mechanism that will be used for such cleanup in various scenarios is not detailed out yet and will be added here before the proposal is complete.

Reviewer 2: ```suggestion

```

> _TODO_: What does the prototype currently do?

Reviewer 2: ```suggestion
```
